JQuery validation for grouped inline field elements

mattpepper opened the pull request:
# Problem

Error handling for grouped inline fields, such as a date group with day:month:year elements, needs to be enable field specific and group level validation messages.
# Issue

Form validation does not support the presentation of group and field specific validation messages on grouped inline field elements. 

At present the validation message specific to the last field:
- is the only message displayed inline when validation occurs;
- is the group error message displayed in the error summary;
## Issue Example
### Current inline date field validation & errors - date of birth

![dates - current inline date field validation & errors - date of birth](https://cloud.githubusercontent.com/assets/5468091/19444247/98d4d072-9487-11e6-8f2b-8c945de3dc56.gif)
# Solution

Add validation for handling grouped inline field elements to display group and field error's inline and group error only in error-summary list. The changes have been UX approved.
- adds 'formInlineGroup' for handling grouped inline form elements (i.e. dates: dob.day dob.month dob.year):
  - merges generic form validation logic and overrides the validation callbacks where grouped elements present;
  - adds groups setting for validator;
  - display error messages inline with group error;
  - display only group error in error summary block;
- adds 'dateGroupValid', specific for _date_ groups, handling inline fields to check valid date is created from day / month / year values entered by user
  - called once rule(s) pass for the grouped field element is the validation context;
  - runs rules for non-context fields in the group to ensure passed;
  - ensures date is a parse-able calendar date (i.e. 30 Feb 2016 or 31 11 2016 is not accepted) 
    - converts 3 character month abbreviations to numeric value
    - converts 2 digit year to 4 digit with logical assumed century
## Solution Examples:
### Inline date field validation & errors - date of birth

![dates - date of birth solution example](https://cloud.githubusercontent.com/assets/5468091/19442516/75d6329c-9481-11e6-9c29-cd35f56bcca4.gif)

rpowis commented:
@mattpepper Wowzer. It's a biggie.

We can review the code and its purpose more thoroughly next week, but for now...

> needs to be improved.
... isn't really a problem statement. Can you expand on what the actual issues were please? Preferably with some images/gifs to highlight demonstrate them clearly?

You don't need to gif the name and NI number inputs in these pages. Can you remove them from the page (using dev tools) before recording so we can focus on the inputs at hand?

Do we also need both the DoB and Passport examples? They don't look any different to me. Only the DoB would do as it's more generally applicable.

rpowis commented on assets/test/specs/fixtures/form-fixture.html:
```diff
@@ -78,6 +78,265 @@ <h2 id="errorSummaryHeading" class="h3-heading">There are errors on the page.</h
     </div>
   </fieldset>
+
+  <fieldset class="form-field-group touch" id="fromDate">
```
This fixture is getting very big. Could probably split this stuff out a bit

rpowis commented on assets/test/specs/fixtures/form-fixture.html:
```diff
@@ -78,6 +78,265 @@ <h2 id="errorSummaryHeading" class="h3-heading">There are errors on the page.</h
+            data-msg-pattern="The ‘no min/max date’ month you have entered is invalid. Enter the month as a number between 1 and 12."
+            data-msg-dategroupvalid="Enter a valid ‘no min/max date’."
+            data-msg-required="Enter the month for the ‘no min/max date’."
+<<<<<<< HEAD
```
🤔

rpowis commented on assets/test/specs/fixtures/form-fixture.html:
```diff
@@ -78,6 +78,265 @@ <h2 id="errorSummaryHeading" class="h3-heading">There are errors on the page.</h
+            data-msg-required="Enter the year for the ‘no min/max date’."
+            data-rule-required="true"
+            data-rule-pattern="^((?:19|20)?\d{2})$"
+<<<<<<< HEAD
```
🤔

rpowis commented on assets/test/specs/form.spec.js:
```diff
@@ -297,21 +342,599 @@ describe('Form Validation', function () {
       });    
     });
+
+    describe('When I use inline group date elements with min and max ‘dateGroupValid’ parameters', function() {
```
Likewise with the fixture, this could probably do with being split out

MarkPaxton commented:
DoB and Passport have different date formats (number, 3 letter month). My service requires full month names, that would be a good addition to this to make it more generic.

feedmypixel commented:
@mattpepper I have a few concerns with this PR.

First and foremost has UX and @hmrc/designers had their eyes on this? Is this functionality the functionality that is wanted? I don't feel it is worth looking at this PR until that has been confirmed. 

IMO there should be only one error message for a group of inputs and a form hint to advise what format the inputs should take. The frontend validation work should then only be concerned with treating a group of inputs as one for its error messages and the summary messages.

Bar that a few concerns are:
- `formInlineGroup.js` feels like an extension when ideally all that is needed is to update `form.js`
- as @rpowis mentioned the fixtures and tests are huge these should be split out to be more manageable
- fixtures should only have what they need in them

But I wouldn't approach any of this work until you have UX sign off

mattpepper commented:
@ben.chidgey - I have UX sign off from David Mcdowell. This requirement has also been raised by another  team, although I am unaware if this has UX sign-off. Perhaps @MarkPaxton can comment.

I am addressing the concerns raised by @rpowis at present, starting with updating the PR description above. I'll then address the splitting out test fixture and spec.

Thanks.

MarkPaxton commented:
Please could you point me to the UX sign off procedure? Thanks

mattpepper commented:
@mark.paxton - the UX sign off procedure for this PR was a demo and 1:1 conversation with the team's UX person.

mattpepper commented:
* PR's Problem/Issue/Solution updated to more accurately reflect current issue, with revised animated GIFs focussing on inline ‘date of birth’ group only 
* Fixture and test for grouped inline elements separated out from other form validation tests/fixture;

@feedmypixel  I haven't addressed the following:
* "formInlineGroup.js feels like an extension when ideally all that is needed is to update form.js"

The approach used was intended to remove possible release breaking issues from the exiting ‘generic’ form validation code. Happy to talk about alternative approaches if this isn't thought to be  a concern.

feedmypixel commented:
@mattpepper for the moment this PR needs to be placed on hold. Discussion needs to occur with the UX and FE members of the teams involved around their date pattern requirements. I will setup a meeting for when your UX is back.

feedmypixel commented:
@mattpepper Re: the meeting around this work please close this PR, and add myself as a watcher on the associated Jira and add the branch associated with this PR to that ticket. 

Please also keep an eye on the [#event-date-pattern](https://hmrcdigital.slack.com/messages/event-date-pattern/details/) channel
